validation/src/blockchain/eth_rpc.rs

```diff
@@ -0,0 +1,295 @@
+use std::{
+    num::NonZero,
+    str::FromStr,
+    sync::{Arc, Mutex},
+};
+
+use anyhow::{anyhow, Context, Result};
+use lru::LruCache;
+use once_cell::sync::Lazy;
+use serde::{Deserialize, Serialize};
+use ssi::caip2;
+use tracing::trace;
+
+const TRANSACTION_CACHE_SIZE: usize = 50;
+const BLOCK_CACHE_SIZE: usize = 50;
+
+static HTTP_CLIENT: Lazy<reqwest::Client> = Lazy::new(reqwest::Client::new);
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct ChainTransaction {
+    /// Transaction hash
+    pub hash: String,
+    /// Transaction contract input
+    pub input: String,
+    pub block: Option<BlockResponse>,
+}
+
+impl ChainTransaction {
+    fn new_from_tx(tx: EthTransaction) -> Self {
+        ChainTransaction {
+            hash: tx.hash,
+            input: tx.input,
+            block: None,
+        }
+    }
+
+    fn try_new_with_block(tx: EthTransaction, block: EthBlock) -> Result<Self> {
+        Ok(ChainTransaction {
+            hash: tx.hash,
+            input: tx.input,
+            block: Some(block.try_into()?),
+        })
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct BlockResponse {
+    pub hash: String,
+    pub number: i64,
+    pub timestamp: i64,
+}
+
+#[async_trait::async_trait]
+pub trait EthRpc {
+    fn chain_id(&self) -> &caip2::ChainId;
+
+    async fn get_block_timestamp(&self, tx_hash: &str) -> Result<Option<ChainTransaction>>;
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct RpcResponse<T> {
+    jsonrpc: String,
+    id: i32,
+    result: Option<T>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+#[serde(rename_all = "camelCase")]
+/// A contract transaction. Expects fields (e.g. input) to exist that aren't required for all transactions.
+pub struct EthTransaction {
+    /// Block hash if mined
+    block_hash: Option<String>,
+    /// Transaction hash
+    hash: String,
+    /// contract input data
+    input: String,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+#[serde(rename_all = "camelCase")]
+pub struct EthBlock {
```

Comment:
some comments on why this is different than BlockResponse and the relationship between those two types would be helpful

Reply:
yea, fair enough. this is the actual RPC response and I wanted the public type the trait relies on to convert the blockchain representation (hex strings) to integers before going up a level

Reply:
yeah makes sense and is what I figured, just would be nice to make it clearer with a couple of comments